docs/source/advanced/loops.rst

```diff
@@ -0,0 +1,399 @@
+            return partial(self._training_step, self._training_step_generator)
+
+        def _get_training_step_generator(self, batch, batch_idx, opt_idx):
+            step_kwargs = _build_training_step_kwargs(
+                self.trainer.lightning_module,
+                self.trainer.optimizers,
+                batch,
+                batch_idx,
+                opt_idx,
+                hiddens=None,
+            )
+
+            # Here we are basically calling lightning_module.training_step()
+            # and this returns a generator!
+            generator = self.trainer.accelerator.training_step(step_kwargs)
+            return generator
+
+        def _training_step(self, training_step_generator):
+            lightning_module = self.trainer.lightning_module
+
+            with self.trainer.profiler.profile("model_forward"):
+                lightning_module._current_fx_name = "training_step"
+                with self.trainer.profiler.profile("training_step"):
+
+                    # Here, instead of calling lightning_module.training_step()
+                    # we call next() on the generator!
+                    training_step_output = next(training_step_generator)
+                    self.trainer.accelerator.post_training_step()
+
+                training_step_output = self.trainer.call_hook("training_step_end", training_step_output)
+                result = ClosureResult.from_training_step_output(training_step_output, self.trainer.accumulate_grad_batches)
+            return result
+
+As we can see, not much work needs to be done to enable our generator training step.
+The new loop is called :code:`YieldLoop` and contains a reference to the generator returned by the :code:`training_step`.
+On every new run (over the optimizers) we call the :code:`training_step` method on the LightningModule which is supposed to return a generator because it contains :code:`yield` statements.
+There must be as many :code:`yield` statements as there are optimizers.
+
+Given this new loop, here is how you connect it to the Trainer:
+
+.. code-block:: python
+
+    model = LitModel()
+    trainer = Trainer()
+
+    yield_loop = YieldLoop()
+    trainer.fit_loop.epoch_loop.batch_loop.connect(optimizer_loop=yield_loop)
+
+    trainer.fit(model)  # runs the new loop!
+
+Note that we need to connect it to the :class:`~pytorch_lightning.loops.batch.training_batch_loop.TrainingBatchLoop` and we are replacing the default optimizer loop that Lightning provides.
+
+
+Persisting the state of loops
+-----------------------------
+
+.. note::
+    This is an experimental feature and is not activated by default.
+    Set the environment variable `PL_FAULT_TOLERANT_TRAINING = 1` to enable saving the progress of loops.
+    Read more about :doc:`fault-tolerant training training <../advanced/fault_tolerant_training>`.
+
+An interesting property of the abstract loop interface is that it can maintain a state.
```

Comment:
```suggestion
The PyTorch Lightning Loop state management design is highly inspired from the PyTorch nn.Module and relies on the same underlying logic. 
```

Reply:
actually all I want to express here is that the benefit of having loops as objects is that they can have state.

Reply:
I know, but I built the state management exactly as pytorch ones. Advanced users are already familiar with it, and it means they will understand faster how it works. Just a personal suggestion.